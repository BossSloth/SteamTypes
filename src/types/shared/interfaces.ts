/* eslint-disable @typescript-eslint/no-explicit-any */
export interface ObservableValue<T> {
  /**
   * Used in {@link Set} to compare the new value with the current value.
   * If it returns true, {@link Set} will not notify subscribers.
   * If this is undefined, {@link Set} will do a strict equality check (===).
   */
  m_fnEquals?(value1: T, value2: T): boolean;

  /**
   * Sets the value of the observable.
   * If the value is different from the current value, it will notify subscribers.
   */
  Set(value: T): void;

  /**
   * Subscribes to changes in the value.
   * @param callback The callback to be called when the value changes.
   * @returns An object with an `Unsubscribe` method to remove the subscription.
   */
  Subscribe(callback: (value: T) => void): { Unsubscribe: () => void; };

  m_callbacks: Callbacks<(value: T) => void>;

  m_currentValue: T;

  /**
   * Returns the number of subscribers from {@link m_callbacks.CountRegistered}.
   */
  SubscriberCount: number;

  /**
   * Returns the current value from {@link m_currentValue}.
   */
  Value: T;
}

// TODO: update converter to also detect this type. Make it default to `<unknown, getType()>`
/**
 * @exported
 */
export interface ReducedValue<TInput, TOutput> {
  /**
   * Creates a new input into the reduced value.
   * @param debugName Optional debug name for this input.
   * @returns An object with `Set` and `Delete` methods to manage this input.
   */
  CreateInput(debugName?: string): {
    Set: (value: TInput) => void;

    Delete: () => void;
  };

  /**
   * The reducer function used to combine all input values into a single output.
   */
  m_fnReducer(inputs: TInput[]): TOutput;

  /**
   * Forces recomputation of the reduced value from all current inputs.
   */
  RecalculateValue(): void;

  /**
   * The internal log instance used for debugging.
   */
  Log: Log;

  /**
   * Internal map of all input values, keyed by their assigned ID.
   */
  m_mapInputs: Map<number, {
    value: TInput;

    strDebugCallstack: string | null;

    strDebugName: string;
  }>;

  /**
   * Incrementing ID counter for assigning keys to inputs.
   */
  m_nNextInputKey: number;

  /**
   * Optional name for debugging/logging.
   */
  m_strNameForLog: string | null;

  /**
   * The underlying observable that stores the reduced value.
   */
  m_SubscribableValue: ObservableValue<TOutput>;

  /**
   * Provides access to the internal subscribable reduced value.
   */
  SubscribableValue: ObservableValue<TOutput>;

  /**
   * Returns the current reduced value.
   */
  Value: TOutput;
}

/**
 * @exported
 */
export interface MappedObservable<TSource, TMapped> {
  /** @native */
  GetValue(): unknown;

  /**
   * The mapping function applied to the source value.
   * Whenever the source changes, this is used to compute the mapped value.
   */
  m_fnMap(value: TSource): TMapped;

  /**
   * Subscribes to changes in the mapped value.
   * @param callback The callback to be called when the mapped value changes.
   * @returns An object with an `Unsubscribe` method to remove the subscription.
   */
  Subscribe(callback: (value: TMapped) => void): { Unsubscribe: () => void; };

  /** @native */
  SyncStore(): unknown;

  /**
   * Forces recomputation of the mapped value from the original source.
   * Also updates the internal {@link m_mappedSubscribableValue}.
   */
  UpdateMappedValue(): void;

  /**
   * Flag indicating whether the mapped value needs to be recomputed.
   * If true, {@link UpdateMappedValue} will be called before returning {@link Value}.
   */
  m_bMappedValueStale: boolean;

  /**
   * The internal observable that stores the mapped value and handles subscriptions.
   */
  m_mappedSubscribableValue: ObservableValue<TMapped>;

  /**
   * The original source observable that this mapped value depends on.
   */
  m_originalSubscribableValue: ObservableValue<TSource>;

  /**
   * Returns the current mapped value.
   * If the value is stale, recomputes it first.
   */
  Value: TMapped;
}

export interface Callbacks<T extends (...args: any) => unknown = () => void> {
  /**
   * Removes all registered callbacks in {@link m_vecCallbacks}.
   */
  ClearAllCallbacks(): void;

  /**
   * @returns the number of registered callbacks in {@link m_vecCallbacks}.
   */
  CountRegistered(): number;

  /**
   * Calls all registered callbacks with the given arguments.
   *
   * @param args The arguments to pass to the callbacks.
   */
  Dispatch(...args: Parameters<T>): void;

  Register(callback: T): { Unregister: () => void; };

  m_vecCallbacks: T[];
}

/**
 * @exported
 */
export interface Unsubscribable {
  /**
   * Unregisters the callback.
   */
  Unregister(): void;
}

export interface Log {
  Assert(): unknown;

  Debug(): unknown;

  Error(): unknown;

  Info(): unknown;

  IsDebugEnabled(): unknown;

  Log(e: unknown, ...t: unknown[]): void;

  m_fnIdGenerator(): unknown;

  Warning(): unknown;

  m_sName: string;
}
