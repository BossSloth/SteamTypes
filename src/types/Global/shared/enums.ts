/** @public */
export enum ESteamUISound {
  LaunchGame,
  FriendMessage,
  ChatMention,
  ChatMessage,
  ToastMessage,
  ToastAchievement,
  ToastMisc,
  ToastMiscShort,
  FriendOnline,
  FriendInGame,
  VolSound,
  ShowModal,
  HideModal,
  IntoGameDetail,
  OutOfGameDetail,
  PagedNavigation,
  ToggleOn,
  ToggleOff,
  SliderUp,
  SliderDown,
  ChangeTabs,
  DefaultOk,
  OpenSideMenu,
  CloseSideMenu,
  BasicNav,
  FailedNav,
  Typing,
  TimerExpired,
  Screenshot,
}
