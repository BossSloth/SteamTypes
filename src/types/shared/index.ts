export * from './enums';
export * from './interfaces';
export * from './protobuf';
export * from './steamid';
