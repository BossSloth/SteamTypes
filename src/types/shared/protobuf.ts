import { ExtensionFieldInfo } from 'google-protobuf';
import { CMsgProtoBufHeader } from '../Protobufs/steam/steammessages_base';

/**
 * To get the deserialized object call {@link ProtobufNotification.Body()}.toObject()
 * @todo: Find way to add this to the compare maps to check for new types
 */
export interface ProtobufNotification<T> {
  BIsValid(): boolean;

  Body(): ProtobufInterface<T>;

  BSuccess(): boolean;

  DEBUG_LogToConsole(): void;

  DEBUG_ToObject(): object | unknown;

  GetEMsg(): number;

  GetEResult(): number;

  GetErrorMessage(): string;

  Hdr(): Header;

  InitForType(e: unknown): void;

  ReadBodyFromBuffer(e: unknown, t: unknown): void;

  Serialize(): Uint8Array;

  SerializeBody(): Uint8Array;

  SetBodyFields(e: unknown): void;

  SetBodyJSON(e: unknown): void;

  SetEMsg(e: number): void;

  m_body: Body;

  m_bValid: boolean;

  m_cubHeader: number;

  /**
   * This value is an enum
   * @currentValue 146
   */
  m_eMsg: number;

  m_header: Header;

  m_netPacket: NetPacket;
}

export type Header = ProtobufInterface<CMsgProtoBufHeader>;

export interface NetPacket {
  GetCountBytesRemaining(): number;

  GetPacket(): unknown;

  GetUint8(): unknown;

  /**
   * @param e default: !0
   */
  GetUint32(e?: boolean): unknown;

  PutBytes(e: unknown): void;

  PutUint8(e: unknown): void;

  /**
   * @param t default: !0
   */
  PutUint32(e: unknown, t?: boolean): void;

  SeekGetCurrent(e: unknown): void;

  /**
   * @param e default: 0
   */
  SeekGetHead(e?: number): void;

  SeekPut(e: unknown): void;

  TellGet(): unknown;

  TellMaxPut(): unknown;

  TellPut(): unknown;

  m_iGet: number;

  m_iPut: number;

  m_nLength: number;

  m_nOffset: number;

  m_rgubPacket: Uint8Array;

  m_viewPacket: DataView;
}

type ProtoGetters<T> = {
  [K in keyof T]-?: () => Exclude<T[K], undefined>;
};

type ProtoSetters<T> = {
  [K in keyof T as `set_${string & K}`]-?: (value: T[K]) => ProtobufInterface<T>;
};

// Helper to check if a type is an interface (object but not array, primitive, or built-in)
type IsInterface<T> = T extends unknown[]
  ? false
  : T extends object
    ? T extends Function
      ? false
      : T extends Date | RegExp | Map<unknown, unknown> | Set<unknown> | WeakMap<object, unknown> | WeakSet<object> | Promise<unknown>
        ? false
        : true
    : false;

// Extract element type from array
type ArrayElement<T> = T extends (infer E)[] ? E : never;

// Conditional type for add method parameter - wraps interfaces in ProtobufInterface
type AddMethodParam<T> = IsInterface<T> extends true ? ProtobufInterface<T> : T;

// Add methods for all array properties
type ProtoAdders<T> = {
  [K in keyof T as Exclude<T[K], undefined> extends unknown[]
    ? `add_${string & K}`
    : never]: (
    value: AddMethodParam<ArrayElement<Exclude<T[K], undefined>>>,
  ) => void;
};

/**
 * Protobuf interface with getters, setters, and add methods for array properties
 * @note You probably don't want to use most of the properties here and
 * instead use the {@link ProtobufInterface.toObject()} method to convert the protobuf message to a plain object
 */
export type ProtobufInterface<T> = ProtoGetters<T> & ProtoSetters<T> & ProtoAdders<T> & SimpleJsPbMessage<T>;

export interface SimpleJsPbMessage<T> {
  clone(): this;

  cloneMessage(): this;

  /**
   * Returns the exact name of the protobuf message "class"
   */
  getClassName(): string;

  getExtension<T>(fieldInfo: ExtensionFieldInfo<T>): T;

  getJsPbMessageId(): unknown | undefined;

  serializeBase64String(): string;

  serializeBinary(): Uint8Array;

  setExtension<T>(fieldInfo: ExtensionFieldInfo<T>, value: T): void;

  toArray(): SimpleJsPbMessage<T>['array'];

  toObject(e?: boolean): T;

  array: InterfaceToTuple<T>;

  arrayIndexOffset_: number;

  /**
   * SimpleJsPbMessage[] of all nested objects in this message
   */
  wrappers_: SimpleJsPbMessage<unknown>[];
}

type InterfaceToTuple<T> = T extends object
  ? {
      [K in keyof T]-?: Exclude<T[K], undefined> extends infer V
        ? V extends object
          ? V extends unknown[]
            ? V // Keep arrays as-is
            : InterfaceToTuple<V> // Recursively convert nested interfaces
          : V
        : never;
    }[keyof T][]
  : T;
