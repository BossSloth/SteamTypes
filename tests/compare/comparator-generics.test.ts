import { createTest } from './shared';
import { genericsCases } from './test-cases';

createTest('Generics Cases', genericsCases);
