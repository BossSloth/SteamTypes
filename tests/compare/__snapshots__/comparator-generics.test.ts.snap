// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Comparator Generics Cases > generic interface type in map mismatch - data 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,7 +1,7 @@
export interface DataStore {
-  data: Map<string, number>;
+  data: Map<string, UserData>;
}

export interface UserData<T extends number> {
  email: string;"
`;

exports[`Comparator Generics Cases > generic interface type in map mismatch - id 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,10 +1,10 @@
export interface DataStore {
  data: Map<string, UserData<number>>;
}

-export interface UserData<T extends string> {
+export interface UserData<T extends number> {
  email: string;

  id: T;

  name: string;"
`;

exports[`Comparator Generics Cases > simple generic mismatch 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,10 +1,10 @@
export interface DataStore {
  data: UserData<number>;
}

-export interface UserData<T extends string> {
+export interface UserData<T extends number> {
  email: string;

  id: T;

  name: string;"
`;
